java.net.SocketException: Protocol family unavailable

Graphouse не запускался в `docker` после обновления. 

Лог в консоли:

```
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)

Graphouse path is: /opt/graphouse
Dec 22, 2017 2:35:30 PM org.springframework.context.annotation.AnnotationConfigApplicationContext prepareRefresh
INFO: Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@3a5ed7a6: startup date [Fri Dec 22 14:35:30 MSK 2017]; root of context hierarchy
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Dec 22, 2017 2:35:32 PM org.springframework.context.annotation.AnnotationConfigApplicationContext refresh
WARNING: Exception encountered during context initialization - cancelling refresh attempt
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'graphouseWebServer' defined in class ru.yandex.market.graphouse.config.ServerConfig: Invocation of init method failed; nested exception is java.net.SocketException: Protocol family unavailable
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:539)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:476)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:303)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:299)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:194)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:755)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:757)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:480)
	at ru.yandex.market.graphouse.GraphouseMain.main(GraphouseMain.java:27)
Caused by: java.net.SocketException: Protocol family unavailable
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:321)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at ru.yandex.market.graphouse.GraphouseWebServer.startServer(GraphouseWebServer.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1702)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1641)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1570)
	... 10 more

Exception in thread "main" org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'graphouseWebServer' defined in class ru.yandex.market.graphouse.config.ServerConfig: Invocation of init method failed; nested exception is java.net.SocketException: Protocol family unavailable
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1574)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:539)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:476)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:303)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:299)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:194)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:755)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:757)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:480)
	at ru.yandex.market.graphouse.GraphouseMain.main(GraphouseMain.java:27)
Caused by: java.net.SocketException: Protocol family unavailable
```

Проблема решилась следующими параметрами окружения:

```
- GH__CACHER__BIND_ADDRESS=0.0.0.0
- GH__HTTP__BIND_ADDRESS=0.0.0.0
```

Или настройками:

```
graphouse.cacher.bind-address=0.0.0.0
graphouse.http.bind-address=0.0.0.0
```

--- a/docker/graphouse/run.sh
+++ b/docker/graphouse/run.sh
@@ -46,7 +46,7 @@ graphouse.clickhouse.query-timeout-seconds=${GH__CLICKHOUSE__QUERY_TIMEOUT_SECON
 graphouse.clickhouse.retention-config=${GH__CLICKHOUSE__RETENTION_CONFIG:=}
 
 #metric server and cacher
-graphouse.cacher.bind-address=${GH__CACHER__BIND_ADDRESS:=::}
+graphouse.cacher.bind-address=${GH__CACHER__BIND_ADDRESS:=}
 graphouse.cacher.port=${GH__CACHER__PORT:=2003}
 graphouse.cacher.threads=${GH__CACHER__THREADS:=100}
 graphouse.cacher.socket-timeout-millis=${GH__CACHER__SOCKET_TIMEOUT_MILLIS:=42000}
@@ -57,7 +57,7 @@ graphouse.cacher.writers-count=${GH__CACHER__WRITERS_COUNT:=2}
 graphouse.cacher.flush-interval-seconds=${GH__CACHER__FLUSH_INTERVAL_SECONDS:=5}
 
 #Http server (metric search, ping, metricData)
-graphouse.http.bind-address=${GH__HTTP__BIND_ADDRESS:=::}
+graphouse.http.bind-address=${GH__HTTP__BIND_ADDRESS:=}
 graphouse.http.port=${GH__HTTP__PORT:=2005}
 graphouse.http.threads=${GH__HTTP__THREADS:=25}
 

--- a/src/main/java/ru/yandex/market/graphouse/GraphouseWebServer.java
+++ b/src/main/java/ru/yandex/market/graphouse/GraphouseWebServer.java
@@ -1,5 +1,6 @@
 package ru.yandex.market.graphouse;
 
+import com.google.common.base.Strings;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.eclipse.jetty.server.Connector;
@@ -25,7 +26,7 @@ public class GraphouseWebServer {
 
     @Value("${graphouse.http.port}")
     private int httpPort;
-    
+
     @Value("${graphouse.http.bind-address}")
     private String httpBindAddress;
 
@@ -54,7 +55,9 @@ public class GraphouseWebServer {
         Server server = new Server(new QueuedThreadPool(threadCount));
         ServerConnector serverConnector = new ServerConnector(server);
         serverConnector.setPort(httpPort);
-        serverConnector.setHost(httpBindAddress);
+        if (!Strings.isNullOrEmpty(httpBindAddress)) {
+            serverConnector.setHost(httpBindAddress);
+        }
         server.setConnectors(new Connector[]{serverConnector});
         ServletContextHandler context = new ServletContextHandler();
         context.setMaxFormContentSize(maxFormContextSizeBytes);

--- a/src/main/java/ru/yandex/market/graphouse/server/MetricServer.java
+++ b/src/main/java/ru/yandex/market/graphouse/server/MetricServer.java
@@ -1,5 +1,6 @@
 package ru.yandex.market.graphouse.server;
 
+import com.google.common.base.Strings;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.springframework.beans.factory.InitializingBean;
@@ -10,10 +11,10 @@ import ru.yandex.market.graphouse.cacher.MetricCacher;
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStreamReader;
+import java.net.InetSocketAddress;
 import java.net.ServerSocket;
 import java.net.Socket;
 import java.net.SocketAddress;
-import java.net.InetSocketAddress;
 import java.net.SocketTimeoutException;
 import java.util.ArrayList;
 import java.util.List;
@@ -31,7 +32,7 @@ public class MetricServer implements InitializingBean {
 
     @Value("${graphouse.cacher.port}")
     private int port;
-    
+
     @Value("${graphouse.cacher.bind-address}")
     private String bindAddress;
 
@@ -60,9 +61,13 @@ public class MetricServer implements InitializingBean {
     public void afterPropertiesSet() throws Exception {
         log.info("Starting metric server on port: " + port);
         serverSocket = new ServerSocket();
-        SocketAddress socketAddress = new InetSocketAddress(bindAddress, port);
+        SocketAddress socketAddress;
+        if (Strings.isNullOrEmpty(bindAddress)) {
+            socketAddress = new InetSocketAddress(port);
+        } else {
+            socketAddress = new InetSocketAddress(bindAddress, port);
+        }
         serverSocket.bind(socketAddress);
-
         log.info("Starting " + threadCount + " metric server threads");
         executorService = Executors.newFixedThreadPool(threadCount);
         for (int i = 0; i < threadCount; i++) {

--- a/src/main/resources/graphouse-default.properties
+++ b/src/main/resources/graphouse-default.properties
@@ -17,7 +17,7 @@ graphouse.clickhouse.query-timeout-seconds=120
 graphouse.clickhouse.retention-config=
 
 #metric server and cacher
-graphouse.cacher.bind-address=::
+graphouse.cacher.bind-address=
 graphouse.cacher.port=2003
 graphouse.cacher.threads=100
 graphouse.cacher.socket-timeout-millis=42000
@@ -31,7 +31,7 @@ graphouse.cacher.max-batch-time-seconds=5
 graphouse.cacher.max-output-threads=5
 
 #Http server (metric search, ping, metricData)
-graphouse.http.bind-address=::
+graphouse.http.bind-address=
 graphouse.http.port=2005
 graphouse.http.threads=25
 graphouse.http.max-form-context-size-bytes=500000

